### ISSUE TITLE
Extra vocab used for tokenization and feature extraction with bert?

### ISSUE THREAD

**Issue by Stuffooh** [jayleicn/TVQAplus#13]
Hi,

I have been working on finetuning and extracting features using the bert model from transformers to try and reproduce the results using the original extracted features. However validation keeps hovering around 50% accuracy no matter what I try.

After some tests I found the following:
```
Key: 103392_q
Model: bert-base-uncased
Sentence: Why is Howard asking questions about the food after Leonard gives him a carton ?

Encoded sentence:
  [101, 2339, 2003, 4922, 4851, 3980, 2055, 1996, 2833, 2044, 7723, 3957, 2032, 1037, 11122, 2239, 1029, 102]

Decoded sentence:
  [CLS] why is howard asking questions about the food after leonard gives him a carton? [SEP]

h5py shape of original extracted features: (15, 768)
shape of new extracted features using default bert tokenization: (16, 768)
103392_q_len in json file: 15
```

When using the default bert tokenizer with the default bert vocab an extra token gets added. This is because the word "carton" is not in the default bert vocab and gets split and tokenized as "cart" and "###on". 

This is the case for a lot of extracted features in the h5py files used by STAGE. The shapes do not match the output of the default bert tokenizer indicating extra tokens/vocab have been added before finetuning/feature extraction. I have used bert-base-cased, bert-base-uncased and bert-large-uncased to see if there are any differences but all output extra tokens compared to the original extracted features.

Could you please share some more information on the feature extraction so me and other researchers in the future can learn from your methods and reproduce your results.

I can give more examples or share the code used to easily compare output if it helps. I'm looking forward to look more into this issue together.

**Comment by jayleicn** [comment]
Hi @Stuffooh, 

I did not use any new vocab. For the words that are not in the vocab, WordPiece Tokenizer will split it into subwords that are in the vocab, e.g., `carton = cart + ###on` as you mentioned. So that you will have two embeddings (`cart` and `###on`) for the word `carton` after feature extraction. To get the embedding for `carton`, I averaged the two embeddings. This might be a critical missing step in your feature extraction pipeline. Sorry that I failed to mention it in #12. The code snippet below shows how this step works, hope it helps.

```python
import numpy as np


def wordpiece_tokenize_with_word_map(wordpiece_tokenizer, tokens):
    """ Further tokenize words into subwords with WordPiece tokenizer.
    Args:
        tokens: list(str), each str is an individual token
        wordpiece_tokenizer: sub-word level tokenizer that splits a word into
            sub-words when possible, e.g. `carton = cart + ###on`,
    Returns:
        wp_tokens: list(str), each str is a sub-word
        wp_token_map: list(int), len(wp_token_map) == len(tokens),
            wp_token_map[i+1] denotes the number of sub-words
            that have been created from tokens[i]. This can be
            used to map the sub-words back to the original tokens.
    """
    assert isinstance(tokens, list)
    wp_tokens = []
    wp_token_map = []
    for t in tokens:
        wp_token_map.append(len(wp_tokens))
        wp_tokens.extend(wordpiece_tokenizer.tokenize(t))
    return wp_tokens, wp_token_map


def get_original_token_embedding(unpadded_wp_embedding, token_map):
    """ sub-word embeddings from the same original token will be
    averaged to get the original token embedding.
    Args:
        unpadded_wp_embedding: (#word_piece_tokens, hsz), torch.FloatTensor
        token_map: list(int),  use to maps the word piece tokens to
        original tokens, see `wordpiece_tokenize_with_word_map`.
    Returns:
    """
    original_token_embedding = [
        unpadded_wp_embedding[token_map[i]:token_map[i+1]].mean(0)
        for i in range(len(token_map)-1)]
    return np.stack(original_token_embedding)  # (#ori_tokens, hsz)
```

**Comment by Stuffooh** [comment]
Hi @jayleicn 

Thanks for the quick reply like always. This looks like the puzzle piece I have been missing these weeks. I will look into it and keep the issue open for now and report back in a few days.

Thanks a lot.

**Comment by Stuffooh** [comment]
Hi @jayleicn 

I implemented the extra step you mentioned but unfortunately I am still not able to reproduce the results. As of now the validation accuracy is around 56% after 14 epochs with a loss of 2.9. (Loss starts at 2.6 at epoch 1.) Training accuracy is around 90% after 14 epochs.

I will describe the steps I have taken to try and reproduce your results:

1. I finetune bert model from transformers usng the following script:
https://github.com/huggingface/transformers/blob/master/examples/run_language_modeling.py

I don't modify the script and I use the format mentioned in #6 and finetune bert-base-uncased with 3 epochs. No tokens get added or whatsoever. 

2. I then use the new pipeline function of transformers:
https://github.com/huggingface/transformers/blob/master/src/transformers/pipelines.py

In combination with the code below to extract the features on the new finetuned bert model:

```
#import libs
import h5py
import json
import argparse
import os
from transformers import pipeline, AutoTokenizer, AutoModel
from tqdm import tqdm
from pathlib import Path
import numpy as np

#required arguments
parser = argparse.ArgumentParser(description='Convert TVQA dataset to new json format with video names as key and sub as value.')

parser.add_argument("--input_path", required=True, type=str, help="Input path for feature extraction pipeline.")
parser.add_argument("--output_path", required=True, type=str, help="Output path for feature extraction pipeline.")

#optional arguments
parser.add_argument("--overwrite_output", action="store_true", default="True", help="Choose if previous output should be overwritten")
parser.add_argument("--model", default="bert-base-cased", type=str, help="The model architecture used for the feature extraction pipeline.")
parser.add_argument("--config", default="bert-base-cased", type=str, help="The config used for the feature extraction pipeline.")
parser.add_argument("--tokenizer", default="bert-base-cased", type=str, help="The tokenizer used for the feature extraction pipeline.")
parser.add_argument("--device", default=-1, type=int, help="ID to determine which device to run pipeline on. -1 is CPU and 0> is GPU.")

args = parser.parse_args()

#make path competible on every OS
args.input_path = Path(args.input_path)
args.output_path = Path(args.output_path)

#initialize model and tokenizer
tokenizer = args.tokenizer
model = args.model
tokenizer = AutoTokenizer.from_pretrained(tokenizer, add_special_tokens=False)
model = AutoModel.from_pretrained(model)
wp_tokens = []
wp_token_map = []


#initialize pipeline
nlp = pipeline('feature-extraction', model=model, config=args.config, tokenizer=tokenizer, device=args.device)

#check if output path is available and delete if requested
if (
   os.path.exists(args.output_path)
   and not args.overwrite_output
):
  raise ValueError(
    "Output path ({}) already exists. Use --overwrite_output or choose a different path.".format(
      args.output_path
    )
  )

elif (
   os.path.exists(args.output_path)
   and args.overwrite_output
):
  os.remove(args.output_path)


#run pipeline and save output in h5py file
with open(str(args.input_path), "r") as input_data:
  input_data = json.load(input_data)

for key, value in tqdm(input_data.items()):
  sentence_list = list(value.split(" "))
  for w in sentence_list:
    wp_token_map.append(len(wp_tokens))
    wp_tokens.extend(tokenizer.tokenize(w))
  wp_token_map.append(len(wp_tokens))
  features = nlp(value)
  features = np.asarray(features, dtype=np.float32)
  if features.shape[1] == 1:
    features = features.reshape(1,768)
  elif features.shape[1] > 1:
    features = np.squeeze(features)
    features = [
      features[wp_token_map[i]:wp_token_map[i+1]].mean(0)
      for i in range(len(wp_token_map)-1)]
    features = np.stack(features)
  if np.isnan(features).any() == True:
    print('This has nan:', key)
  if np.isinf(features).any() == True:
    print('This has inf:', key)
  h5f = h5py.File(args.output_path, 'a')
  h5f.create_dataset(key, data=features)
  h5f.close()
  wp_tokens = []
  wp_token_map = []
```
This results in a h5py file with the exact same key names and shape for every dataset for the questions and answers.

3. I run "bash run_main.sh --add_local" leaving everything identical except for the QA h5py file which I point to newly created one with the new extracted features. 

At this point I'm a little at a loss why the accuracy is so low. I can understand if the accuracy would be perhaps 1-2% lower because of seeding, slight preprocesing differences and what not but the difference is too high considering I'm reproducing your steps with the same data. I also find it interesting how the validation loss only gets higher. 

Am I missing some important details or steps which explain the poor results? I would be very grateful if you could take another look at things together with me to figure out the big difference in our results.

**Comment by jayleicn** [comment]
Hi @Stuffooh, 

Have you confirmed that you can reproduce the results using the released features? This helps to make sure there is nothing with the environment (PyTorch version, Python version, etc).

Jie

**Comment by Stuffooh** [comment]
Hi @jayleicn 

Yes when using your extracted features I can reproduce the results with no problem. But when using my own extracted features the results are very different.

**Comment by Stuffooh** [comment]
I have done some more testing and digging and found the following:

1. When I use the original features I can reproduce the results.
2. When I use the original QA features and my own extracted sub features I can reproduce the results.
3. When I use my own extracted QA features and the original sub features I can not reproduce the results (~52%).
4.  When I use my own extracted QA and sub features I can not reproduce the results (~52%).

I took a good look at the difference between the original QA features and my own extracted features. The amount of datasets, name of datasets, shape and data types are identical. However the HDF5 file of my extracted QA features are 3.8GB while the original is 3.6GB. The sub are identical in size however. So for some reason my extracted QA feature HDF5 file is bigger in size and causes for poor performance during training/evaluation. I cannot come up with an explanation for this other than maybe there is a difference in how the QA and sub features get extracted and/or the numpy arrays get manipulated (compressed/shortened) in some way before being saved to the HDF5 datasets.

Anyway I reached a point where I will halt my attempts to reproduce the results. I hope other people may find some of the information helpful to reproduce or improve the results in the future.

If at some point someone gets it working please let me know.

**Comment by jayleicn** [comment]
Hi @Stuffooh,

Thanks for your detailed testing! 

After a detailed look at your code, I found some noticeable differences with my original script.
- I used "bert-base-uncased" instead of "bert-base-cased". The uncased version would give you a smaller set of vocab, which might help the convergence of the model.
- I added special_tokens in both fine-tuning and feature extraction. They are removed at the embedding level, i.e., "Sheldon is eating an apple" would become something similar to "[CLS] Sheldon is eating an apple [SEP]". The later one is sent to the fine-tuned model to extract a `E = 7x768` embedding. The next step is to remove the embedding for [CLS] and [SEP],  the final embedding is `E_{final} = E[1:6] = 5x768`.
- The original features are the second-to-last layer hidden states of the model (as stated in Sec. 4.2 of the TVQA+ paper), in contrast, `nlp = pipeline('feature-extraction', model=model, config=args.config, tokenizer=tokenizer, device=args.device)` extracts the last layer hidden states. The comparison on CoNLL-2003 NER task suggests that the hidden states from the second-to-last layer are better than that from the last, ref: https://arxiv.org/pdf/1810.04805.pdf, Table 7.

I hope these details would help.
